### PR TITLE
feat(edgeless): support top-level image block inside surface

### LINKED ISSUE #3594: Support top-level image block in whiteboard
The current images must appear in the Note component. It is hoped that images can be inserted independently on the whiteboard. In addition, based on this capability, it is supported to select from the sticker library provided by the design Team, or insert from some common external resources, such as Unsplash.

### PR DESCRIPTION
**Pull request by regischen**
close #3594

### CODE REVIEW

**Comment by Flrande** [comment]
image

File: packages/blocks/src/image-block/image-block.ts [hunk 1]
```diff
@@ -19,8 +23,157 @@ import { type ImageBlockModel, ImageBlockSchema } from './image-model.js';
...
+          this._fetchError(new Error('Cannot find blob'));
+        }
+      })
+      .catch(this._fetchError);
+  };
+
+  override connectedCallback() {
+    super.connectedCallback();
+    this._fetchImage();
+    this._disposables.add;
```
Comment (lawvs) [inline packages/blocks/src/image-block/image-block.ts:148]:
```suggestion
```

**Review by doodlewind — COMMENTED** [review]
The image dragged from native OS window is still inserted with the note block wrapper.

https://github.com/toeverything/blocksuite/assets/7312949/e9a866e8-e863-4ef8-87de-b95e38d5d820

File: packages/blocks/src/page-block/edgeless/edgeless-blocks-container.ts [hunk 2]
```diff
@@ -166,11 +175,20 @@ export class EdgelessBlockContainer extends WithDisposable(LitElement) {
...
-          <edgeless-notes-container
-            .edgeless=${edgeless}
-            .notes=${sortedNotes}
-            .renderer=${renderModel.bind(this)}
-          ></edgeless-notes-container>
+          ${repeat(
+            blocks,
+            block => block.id,
+            (block, index) => {
+              const tag = literal`${unsafeStatic(
```
Comment (doodlewind) [inline packages/blocks/src/page-block/edgeless/edgeless-blocks-container.ts:182]: The block flavour here comes from unreliable dynamic content, so using it transparently as HTML tag type seems to be unsafe.

File: packages/blocks/src/image-block/image-block.ts [hunk 3]
```diff
@@ -19,8 +23,157 @@ import { type ImageBlockModel, ImageBlockSchema } from './image-model.js';
...
+        this._fetchImage();
+        // 1s, 2s, 3s
+      }, 1000 * this._retryCount);
+    } else {
+      console.error(e);
+      this._imageState = 'failed';
+    }
+  };
+
+  protected _fetchImage = () => {
```
Comment (doodlewind) [inline packages/blocks/src/image-block/image-block.ts:119]: Why `protected`? Are these protected methods designed to be inherited?

File: packages/blocks/src/image-block/image-block.ts [hunk 4]
```diff
@@ -19,8 +23,157 @@ import { type ImageBlockModel, ImageBlockSchema } from './image-model.js';
...
+        .root=${this.root}
+        .widgets=${this.widgets}
+        .content=${this.content}
+      ></affine-edgeless-image>`;
+    }
+  }
+}
+
+class ImageBlock extends BlockElement<ImageBlockModel> {
+  static maxRetryCount = 1;
```
Comment (doodlewind) [inline packages/blocks/src/image-block/image-block.ts:75]: Define this as `MAX_RETRY_COUNT` at the top of this module.

File: packages/blocks/src/surface-block/surface-block.ts [hunk 5]
```diff
@@ -265,6 +315,222 @@ export class SurfaceBlockComponent extends BlockElement<SurfaceBlockModel> {
...
+
+    this._initBlocks();
+  }
+
+  private _initBlocks() {
+    const { page } = this;
+    this.blocks.forEach(block => {
+      this._addToBatch(block);
+    });
+    page.slots.blockUpdated.on(e => {
```
Comment (doodlewind) [inline packages/blocks/src/surface-block/surface-block.ts:327]: Why this is not wrapped in `this._disposables.add()`?

File: packages/blocks/src/image-block/image-model.ts [hunk 6]
```diff
@@ -24,6 +40,33 @@ export const ImageBlockSchema = defineBlockSchema({
     role: 'content',
   },
   transformer: () => new ImageBlockTransformer(),
+  toModel: () => {
+    return new ImageBlockModel();
+  },
 });
 
-export type ImageBlockModel = SchemaToModel<typeof ImageBlockSchema>;
+@RectElement
```
Comment (doodlewind) [inline packages/blocks/src/image-block/image-model.ts:48]: This decorator duplicates the concept of *element*. Is it indicating top-level content in edgeless mode? If so, it's also weird that shape, connector and frame are not decorated by it. Could you elaborate?

File: packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts [hunk 7]
```diff
@@ -184,8 +238,29 @@ export class EdgelessClipboard implements Clipboard {
...
+      sourceIds: string[];
+      images: string[];
+    } = {
+      data: data.data,
+      sourceIds: data.sourceIds,
+      images: [],
+    };
+    if (data.imageBlobs.length > 0) {
+      copyData.images = await Promise.all(
+        data.imageBlobs.map(blob => blobToBase64(blob))
```
[the remark this answers is not in the document]
Comment (regischen) [reply packages/blocks/src/__internal__/clipboard/edgeless-clipboard.ts]: this was intended to fix #4967 , but never mind, I am intented leave it to another pr.
In short, I find this way to copy multiple images. 
no, the base64 doesn't exist in the block tree.

File: packages/blocks/src/image-block/image-model.ts (hunk 6 above)
Comment (regischen) [reply packages/blocks/src/image-block/image-model.ts]: It is Mixin, some utils of rect-like. so changing a name would be great.

File: packages/blocks/src/page-block/edgeless/edgeless-blocks-container.ts (hunk 2 above)
Comment (regischen) [reply packages/blocks/src/page-block/edgeless/edgeless-blocks-container.ts:212]: in this case, the flavour is one of note or image, so I think it's safe.

File: packages/blocks/src/page-block/edgeless/components/image/edgeless-image.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import { LitElement } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { styleMap } from 'lit/directives/style-map.js';
+import { html } from 'lit/static-html.js';
+
+import type { ImageBlockModel } from '../../../../image-block/index.js';
+import { Bound } from '../../../../surface-block/index.js';
+import type { SurfaceBlockComponent } from '../../../../surface-block/surface-block.js';
+
+@customElement('edgeless-image')
```
Comment (doodlewind) [inline packages/blocks/src/page-block/edgeless/components/image/edgeless-image.ts:11]: Now we have  both `edgeless-image` and `affine-edgeless-image` in the codebase, which is duplicate. How to differentiate them?

**Review by doodlewind — APPROVED** [review]
Should land this after the cleanups, thanks for the feat.

**Pull request merged**